Infra

paul-khudan opened the pull request:
Suggestion to add Task and Runner interfaces.
1 Task defines a pipeline step interface.
2 Runners are used to configure, prepare and run tasks from different environments. We add a default implementation of SandboxRunner - it runs a task defined in a python module file in an isolated filesystem sandbox. Runner interface makes it easy to implement various running, e.g. from command line:
`python run.py --task configs/ukr/train-tokenizer/ukr-roberta-base.py`

Task now is identified with a sequence of abstract `str` identifiers - there is no concrete types like `language`, `name`, `task`. For SandboxRunner , these identifiers are defined as a path to the config file (except default `config` folder name)

Commit:
update requirements

Commit:
introduce Task, Sandbox and Runner interfaces. Move legacy code to new infra

Commit:
correct configs (input paths)

Commit:
rename argument

Commit:
back to double-quotes

EvGe22 commented on .pre-commit-config.yaml:
```diff
@@ -20,7 +21,17 @@ repos:
       - id: check-json
       - id: end-of-file-fixer
       - id: requirements-txt-fixer
+  - repo: https://github.com/pycqa/flake8
```
Why not https://github.com/youscan/python-codestyle ?

paul-khudan replied:
Mainly for personal development reasons - I cannot run locally the provided docker image, so I updated the current local configuration.
This could be excluded from vcs, to keep the functionality local.

Commit:
fix output folder use. remove debug class

EvGe22 commented on src/language_model/runner.py:
```diff
@@ -0,0 +1,26 @@
+
+
+class ConfigurationFileRunner(TaskRunner):
+    def __init__(self, config_path: str):
+        self.config_path = config_path
+        module_name = os.path.splitext(config_path)[0].replace("/", ".")
+        module = import_module(module_name)
+        self.task: SandboxTask = getattr(module, TASK_FIELD_NAME)
```
I'd suggest moving this to the run method or create another method e.g. `run_from_config`. 
In my understanding, the `__init__` method of the `ConfigurationFileRunner` should configure the runner itself. I.e. number of threads to use, logging, multiprocessing, etc...

paul-khudan replied:
Following the "fail fast" concept, we initialize the maximum amount of variables as possible in the constructor method.
I see that part of initialization is implemented in `run` method. Need to review that to move everything to `__init__`.
What are the drawbacks of such an approach?

EvGe22 replied:
The only drawback I see is that I find this combination of naming-behavior a bit counterintuitive. The `Runner` class to me sounds like a `Worker` in Celery - a thing that is created once and that accepts different tasks to run, the `Runner` initialization in this case should configure *how* we run the tasks, not which tasks we run. 
In the current implementation, the Runner is created per task and its task is set in the `__init__`. 

This class is responsible for obtaining the task via a path and supping it with essentially an output folder. Not exactly sure the `Runner` is a correct abstraction name here. I see it like we have a `TaskDescription` with the configuration and the commands that should be run to complete this task (`class Task` with `def execute`). And this `Runner` is more like a `TaskInstance` that should be run with a specific output folder. I do not really like the proposed naming, it's here to better convey the message.

I do see what you mean though and since we only run tasks one at a time and mostly via a terminal, this should work just fine.

EvGe22 commented on src/language_model/tokenization/trainer.py:
```diff
@@ -0,0 +1,42 @@
+
+        self.tokenizer.save(os.path.join(environment_path, "tokenizer"))
+
+    @staticmethod
+    def get_all_files_in_folder(data_folder_path: str) -> List[str]:
+        data_files_paths = []
+        for (dir_path, _, filenames) in os.walk(data_folder_path):
+            data_files_paths.extend([os.path.join(dir_path, file_name) for file_name in filenames])
+        return data_files_paths
```
Should this be a separate method? Is there a specific reason it has to be a `@staticmethod` of the `ByteLevelBPETokenizerTrainer` class?

paul-khudan replied:
This is done to conserve an utility method alongside with the class that uses it to easily remove as deprecated in later versions.

EvGe22 commented on src/language_model/pipeline.py:
```diff
@@ -0,0 +1,76 @@
+    return configure_logger(log_path, file=DEFAULT_LOG_FILE, overwrite=overwrite, log_to_stderr=log_to_stderr)
+
+
+def configure_logger(
```
IMO the name implies that the method should configure an existing logger, not create one.

EvGe22 commented on src/language_model/pipeline.py:
```diff
@@ -0,0 +1,76 @@
+        super().__init__(sandbox_folder_path)
+
+
+def identifiers_from_config_file(filepath: str) -> Sequence[str]:
```
The method looks ambiguous:
-  What are identifiers? 
- imo it would be better to name it `*_from_config_path`.

paul-khudan replied:
Each experiment step has a set of identifiers that exhaustively describe it - language, step name, version, etc. We use an ordered `str` sequence to not fixate on some particular identifiers.

EvGe22 commented on src/language_model/pipeline.py:
```diff
@@ -0,0 +1,76 @@
+        return self.sandbox_folder_path
+
+
+class SandboxTask(ITask, AbstractSandbox, ABC):
```
Why `Sandbox`? This looks like running tasks that extend this class will not actually save/do/process anything, but rather test something.

paul-khudan replied:
Do you associate the `sandbox` term with testing only? If this association is strong we would want to find another name for a place where usually experiments are run **in isolation** - I followed this semantics when named it that way.

paul-khudan replied:
I guess, the fact that sandboxes are used for testing, does not make them test utilities only. The concept must be more generic, afaic

paul-khudan replied:
I agree that this sandbox only set the usage of filesystem (pointing to a directory), but not the whole system (devices, ram, etc.)

EvGe22 replied:
My association is the following: Sandbox - some isolated new environment used mostly for harmlessly testing things that might otherwise influence or irreversibly change some parts of the existing environment which we want to stay the same. This Sandbox can easily be destroyed without leaving a trace of its existence apart from maybe some logs.

Commit:
create sandbox folder before running

Commit:
refactor pipeline. resolve task <-> sandbox relationship

# Conflicts:
#	src/language_model/data/extract.py
#	src/language_model/data/load.py
#	src/language_model/modelling/trainer.py
#	src/language_model/tokenization/extend.py
#	src/language_model/tokenization/tasks.py
#	src/language_model/tokenization/trainer.py

Commit:
combine two logger-initting methods
